src/roles.c

```diff
@@ -0,0 +1,525 @@
+	 * offline nodes to spare. */
+	for (i = 0; i < n_cluster; i += 1) {
+		if (!cluster[i].online && cluster[i].role != DQLITE_SPARE) {
+			cb(cluster[i].id, DQLITE_SPARE, arg);
+			cluster[i].role = DQLITE_SPARE;
+		} else if (cluster[i].online &&
+			   cluster[i].role == DQLITE_VOTER) {
+			voter_count += 1;
+			addDomain(cluster[i].failure_domain, &voter_compare);
+		} else if (cluster[i].online &&
+			   cluster[i].role == DQLITE_STANDBY) {
+			standby_count += 1;
+			addDomain(cluster[i].failure_domain, &standby_compare);
+		}
+	}
+
+	/* If we don't have enough voters, promote some standbys and spares. */
+	if (voter_count < voters) {
+		qsort_r(cluster, n_cluster, sizeof *cluster,
+			compareNodesForPromotion, &voter_compare);
+	}
+	for (i = 0; i < n_cluster && voter_count < voters; i += 1) {
+		if (!cluster[i].online || cluster[i].role == DQLITE_VOTER) {
+			continue;
+		}
+		cb(cluster[i].id, DQLITE_VOTER, arg);
+		if (cluster[i].role == DQLITE_STANDBY) {
+			standby_count -= 1;
+			removeDomain(cluster[i].failure_domain,
+				     &standby_compare);
+		}
+		cluster[i].role = DQLITE_VOTER;
+		voter_count += 1;
+		addDomain(cluster[i].failure_domain, &voter_compare);
+	}
+
+	/* If we have too many voters, demote some of them. We always demote
+	 * to spare in this step -- if it turns out that it would be better
+	 * for some of these nodes to end up as standbys, that change will
+	 * be picked up in the next step, and the two role changes will be
+	 * consolidated by queueChangeCb. */
+	if (voter_count > voters) {
+		qsort_r(cluster, n_cluster, sizeof *cluster,
+			compareNodesForDemotion, &voter_compare);
+	}
+	for (i = 0; i < n_cluster && voter_count > voters; i += 1) {
+		if (cluster[i].role != DQLITE_VOTER || cluster[i].id == my_id) {
+			continue;
+		}
+		cb(cluster[i].id, DQLITE_SPARE, arg);
+		cluster[i].role = DQLITE_SPARE;
+		voter_count -= 1;
+		removeDomain(cluster[i].failure_domain, &voter_compare);
+	}
+
+	/* If we don't have enough standbys, promote some spares. */
+	if (standby_count < standbys) {
+		qsort_r(cluster, n_cluster, sizeof *cluster,
+			compareNodesForPromotion, &standby_compare);
+	}
+	for (i = 0; i < n_cluster && standby_count < standbys; i += 1) {
+		if (!cluster[i].online || cluster[i].role != DQLITE_SPARE) {
+			continue;
+		}
+		cb(cluster[i].id, DQLITE_STANDBY, arg);
+		cluster[i].role = DQLITE_STANDBY;
+		standby_count += 1;
+		addDomain(cluster[i].failure_domain, &standby_compare);
+	}
+
+	/* If we have too many standbys, demote some of them. */
+	if (standby_count > standbys) {
+		qsort_r(cluster, n_cluster, sizeof *cluster,
+			compareNodesForDemotion, &standby_compare);
+	}
+	for (i = 0; i < n_cluster && standby_count > standbys; i += 1) {
+		if (cluster[i].role != DQLITE_STANDBY) {
+			continue;
+		}
+		cb(cluster[i].id, DQLITE_SPARE, arg);
+		cluster[i].role = DQLITE_SPARE;
+		standby_count -= 1;
+		removeDomain(cluster[i].failure_domain, &standby_compare);
+	}
+}
```

Comment:
I don't have time right now to go through this in detail, but I assume it's basically a verbatim translation of the equivalent code in `go-dqlite`? In app.RoleChanges.Adjust().